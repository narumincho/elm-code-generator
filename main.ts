import * as data from "./data";
import { strict } from "assert";

const reservedWord = new Set([
  "if",
  "then",
  "else",
  "case",
  "of",
  "let",
  "in",
  "type",
  "module",
  "where",
  "import",
  "port",
  "exposing",
  "as",
  "alias",
]);

const indent = "    ";

export const fieldNameFromString = (
  fieldName: string
): data.Maybe<data.FieldName> => {
  if (/^[a-z][a-zA-Z0-9_]*/u.test(fieldName) && !reservedWord.has(fieldName)) {
    return data.Maybe.Just(data.FieldName.FieldName(fieldName));
  }
  return data.Maybe.Nothing();
};

export const fieldNameFromStringOrThrow = (
  fieldName: string
): data.FieldName => {
  const result = fieldNameFromString(fieldName);
  switch (result._) {
    case "Just":
      return result.value;
    case "Nothing":
      throw new Error("invalid field name = " + fieldName);
  }
};

export const variantNameFormString = (
  variantName: string
): data.Maybe<data.VariantName> => {
  if (
    /^[A-Z][a-zA-Z0-9_]*/u.test(variantName) &&
    !reservedWord.has(variantName)
  ) {
    return data.Maybe.Just(data.VariantName.VariantName(variantName));
  }
  return data.Maybe.Nothing();
};

/**
 * @throws バリアント名にできない名前だった場合
 */
export const variantNameFormStringOrThrow = (
  variantName: string
): data.VariantName => {
  const result = variantNameFormString(variantName);
  switch (result._) {
    case "Just":
      return result.value;
    case "Nothing":
      throw new Error("invalid variant name = " + variantName);
  }
};

export const elmTypeNameFromString = (
  typeName: string
): data.Maybe<data.ElmTypeName> => {
  if (/^[A-Z][a-zA-Z0-9_]*$/u.test(typeName) && !reservedWord.has(typeName)) {
    return data.Maybe.Just(data.ElmTypeName.ElmTypeName(typeName));
  }
  return data.Maybe.Nothing();
};

/**
 * @throws 型名にできない名前だった場合
 */
export const elmTypeNameFromStringOrThrow = (
  typeName: string
): data.ElmTypeName => {
  const result = elmTypeNameFromString(typeName);
  switch (result._) {
    case "Just":
      return result.value;
    case "Nothing":
      throw new Error("invalid elm type name = " + typeName);
  }
};

export const codeToString = (elmCode: data.Code): string => {
  return (
    "module " +
    elmCode.moduleName +
    " exposing (" +
    elmCode.typeDeclarationList.map(typeDeclarationToExportName).join(", ") +
    ")" +
    "\n\n" +
    elmCode.typeDeclarationList.map(typeDeclarationToString).join("\n\n") +
    "\n"
  );
};

const typeDeclarationToExportName = (
  typeDeclaration: data.TypeDeclaration
): string => {
  switch (typeDeclaration._) {
    case "CustomType":
      return customTypeToExportName(typeDeclaration.customType);
    case "TypeAlias":
      if (typeDeclaration.typeAlias.export) {
        return typeDeclaration.typeAlias.name.string;
      }
      return "";
  }
};

const customTypeToExportName = (customType: data.CustomType): string => {
  switch (customType.export) {
    case "NoExport":
      return "";
    case "ExportTypeOnly":
      return customType.name.string;
    case "ExportTypeAndVariant":
      return customType.name.string + "(..)";
  }
};

const typeDeclarationToString = (
  typeDeclaration: data.TypeDeclaration
): string => {
  switch (typeDeclaration._) {
    case "TypeAlias":
      return typeAliasToString(typeDeclaration.typeAlias);
    case "CustomType":
      return customTypeToString(typeDeclaration.customType);
  }
};

const typeAliasToString = (typeAlias: data.TypeAlias): string =>
  "type alias " +
  typeAlias.name.string +
  " =\n" +
  indent +
  elmTypeToString(typeAlias.type);

const fieldToString = (field: data.Field): string =>
  field.name.string + " : " + elmTypeToString(field.type);

const customTypeToString = (customType: data.CustomType): string =>
  "type " +
  customType.name.string +
  "\n" +
  indent +
  "= " +
  customType.variantList.map(variantToString).join("\n" + indent + "| ");

const variantToString = (variant: data.Variant): string =>
  variant.name.string +
  (variant.parameter.length === 0
    ? ""
    : " " + variant.parameter.map(elmTypeToString).join(" "));

const elmTypeToString = (elmType: data.ElmType): string => {
  switch (elmType._) {
    case "LocalType":
      if (elmType.localType.parameter.length === 0) {
        return elmType.localType.typeName.string;
      }
      return (
        "(" +
        elmType.localType.typeName.string +
        " " +
        elmType.localType.parameter.map(elmTypeToString).join(" ") +
        ")"
      );
    case "ImportedType":
      if (elmType.importedType.parameter.length === 0) {
        return (
          elmType.importedType.moduleName +
          "." +
          elmType.importedType.typeName.string
        );
      }
      return (
        "(" +
        elmType.importedType.moduleName +
        "." +
        elmType.importedType.typeName.string +
        " " +
        elmType.importedType.parameter.map(elmTypeToString).join(" ") +
        ")"
      );
    case "Function":
      return (
        "(" +
        elmTypeToString(elmType.functionType.input) +
        " -> " +
        elmTypeToString(elmType.functionType.output) +
        ")"
      );
    case "Tuple0":
      return "()";
    case "Tuple2":
      return (
        "(" +
        elmTypeToString(elmType.tuple2.first) +
        ", " +
        elmTypeToString(elmType.tuple2.second) +
        ")"
      );
    case "Tuple3":
      return (
        "(" +
        elmTypeToString(elmType.tuple3.first) +
        ", " +
        elmTypeToString(elmType.tuple3.second) +
        ", " +
        elmTypeToString(elmType.tuple3.third) +
        ")"
      );
    case "Record":
      return (
        "{ " +
        elmType.fieldList.map((e): string => fieldToString(e)).join(", ") +
        " }"
      );
  }
};
