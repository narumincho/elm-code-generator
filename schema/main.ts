import * as codeGen from "js-ts-code-generator";
import * as coreTypePartMap from "definy-core/schema/typePartMap";
import * as definyCore from "definy-core";
import * as fs from "fs";
import * as prettier from "prettier";
import * as typePartMap from "./typePartMap";

const typeScriptCode = codeGen.generateCodeAsString(
  definyCore.generateTypeScriptCode(
    new Map([...coreTypePartMap.typePartMap, ...typePartMap.typePartMap])
  ),
  "TypeScript"
);

const typeScriptPath = "source/data.ts";
fs.promises
  .writeFile(
    typeScriptPath,
    prettier.format(typeScriptCode, { parser: "typescript" })
  )
  .then(() => {
    console.log("output TypeScript code!");
  });
