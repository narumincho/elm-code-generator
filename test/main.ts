import * as data from "../data";
import * as main from "../main";

describe("test", () => {
  it("type name validation valid", () => {
    expect(main.elmTypeNameFromString("Data")).toEqual<
      data.Maybe<data.ElmTypeName>
    >(data.Maybe.Just(data.ElmTypeName.ElmTypeName("Data")));
  });

  it("type name validation invalid", () => {
    expect(main.elmTypeNameFromString("それな")).toEqual<
      data.Maybe<data.ElmTypeName>
    >(data.Maybe.Nothing());
  });

  it("output variant", () => {
    const colorTypeName = main.elmTypeNameFromString("Color");
    switch (colorTypeName._) {
      case "Just": {
        const sampleElmCode: data.Code = {
          moduleName: "Main",
          typeDeclarationList: [
            data.TypeDeclaration.CustomType({
              name: colorTypeName.value,
              export: data.CustomTypeExportLevel.ExportTypeAndVariant,
              comment: "色",
              variantList: [
                {
                  name: main.variantNameFormStringOrThrow("Red"),
                  parameter: [],
                },
                {
                  name: main.variantNameFormStringOrThrow("Green"),
                  parameter: [],
                },
                {
                  name: main.variantNameFormStringOrThrow("Blue"),
                  parameter: [],
                },
              ],
            }),
          ],
        };
        expect(main.codeToString(sampleElmCode)).toMatchSnapshot();
        return;
      }
      case "Nothing":
        fail();
    }
  });

  it("output type alias", () => {
    const sampleElmCode: data.Code = {
      moduleName: "Main",
      typeDeclarationList: [
        data.TypeDeclaration.TypeAlias({
          name: main.elmTypeNameFromStringOrThrow("User"),
          export: false,
          comment: "色",
          fieldList: [
            {
              name: main.fieldNameFromStringOrThrow("name"),
              type: data.ElmType.ImportedType({
                moduleName: "String",
                typeName: main.elmTypeNameFromStringOrThrow("String"),
              }),
            },
            {
              name: main.fieldNameFromStringOrThrow("age"),
              type: data.ElmType.ImportedType({
                moduleName: "Basics",
                typeName: main.elmTypeNameFromStringOrThrow("Int"),
              }),
            },
          ],
        }),
      ],
    };
    expect(main.codeToString(sampleElmCode)).toMatchSnapshot();
  });

  it("invalid filed name", () => {
    expect(() => {
      const sampleElmCode: data.Code = {
        moduleName: "Main",
        typeDeclarationList: [
          data.TypeDeclaration.TypeAlias({
            name: main.elmTypeNameFromStringOrThrow("IncludeInvalidFiledName"),
            export: false,
            comment: "",
            fieldList: [
              {
                name: main.fieldNameFromStringOrThrow("then"),
                type: data.ElmType.ImportedType({
                  moduleName: "String",
                  typeName: main.elmTypeNameFromStringOrThrow("String"),
                }),
              },
            ],
          }),
        ],
      };

      main.codeToString(sampleElmCode);
    }).toThrowErrorMatchingSnapshot();
  });
});
