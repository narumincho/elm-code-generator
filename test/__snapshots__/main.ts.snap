// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test invalid filed name 1`] = `"invalid field name = then"`;

exports[`test output type alias 1`] = `
"module Main exposing ()

type alias User =
    { name : String.String
    , age : Basics.Int
    }
"
`;

exports[`test output variant 1`] = `
"module Main exposing (Color(..))

type Color
    = Red
    | Green
    | Blue
"
`;
